library/core/src/num/dec2flt/parse.rs

```diff
@@ -1,121 +1,246 @@
+/// Parse up to 19 digits (the max that can be stored in a 64-bit integer).
+fn try_parse_19digits(s: &mut AsciiStr<'_>, x: &mut u64) {
+    // SAFETY: safe since s cannot be empty when first and step are called.
+    unsafe {
+        while *x < MIN_19DIGIT_INT && !s.is_empty() && s.first_unchecked().is_ascii_digit() {
+            let digit = s.first_unchecked() - b'0';
+            *x = (*x * 10) + digit as u64; // no overflows here
+            s.step();
+        }
     }
 }
 
-#[derive(Debug, PartialEq, Eq)]
-pub enum ParseResult<'a> {
-    Valid(Decimal<'a>),
-    ShortcutToInf,
-    ShortcutToZero,
-    Invalid,
+/// Try to parse 8 digits at a time, using an optimized algorithm.
+fn try_parse_8digits(s: &mut AsciiStr<'_>, x: &mut u64) {
+    // may cause overflows, to be handled later
+    if let Some(v) = s.read_u64() {
+        if is_8digits(v) {
+            *x = x.wrapping_mul(1_0000_0000).wrapping_add(parse_8digits(v));
+            // SAFETY: already ensured the buffer was >= 8 bytes in read_u64.
+            unsafe {
+                s.step_by(8);
+            }
+            if let Some(v) = s.read_u64() {
+                if is_8digits(v) {
+                    *x = x.wrapping_mul(1_0000_0000).wrapping_add(parse_8digits(v));
+                    // SAFETY: already ensured the buffer was >= 8 bytes in try_read_u64.
+                    unsafe {
+                        s.step_by(8);
+                    }
+                }
+            }
+        }
+    }
 }
 
-/// Checks if the input string is a valid floating point number and if so, locate the integral
-/// part, the fractional part, and the exponent in it. Does not handle signs.
-pub fn parse_decimal(s: &str) -> ParseResult<'_> {
-    if s.is_empty() {
-        return Invalid;
+/// Parse the scientific notation component of a float.
+fn parse_scientific(s: &mut AsciiStr<'_>) -> i64 {
+    let start = *s;
+    // SAFETY: the first character is 'e'/'E' and scientific mode is enabled
+    unsafe {
+        s.step();
+    }
+    let mut exp_num = 0_i64;
+    let mut neg_exp = false;
+    if s.first_is2(b'-', b'+') {
+        neg_exp = s.first_is(b'-');
+        // SAFETY: s cannot be empty
+        unsafe {
+            s.step();
+        }
+    }
+    if s.first_isdigit() {
+        s.parse_digits(|digit| {
+            // no overflows here, saturate well before overflow
+            if exp_num < 0x10000 {
+                exp_num = 10 * exp_num + digit as i64;
+            }
+        });
+        if neg_exp { -exp_num } else { exp_num }
+    } else {
+        // ignore 'e' and return back
+        *s = start;
+        0
     }
+}
 
-    let s = s.as_bytes();
-    let (integral, s) = eat_digits(s);
+/// Try to parse a non-special floating point number.
+///
+/// This creates a representation of the float as the
+/// significant digits and the decimal exponent.
+pub fn parse_number(s: &[u8]) -> Option<(Number, usize)> {
+    debug_assert!(!s.is_empty());
```

Comment:
You rely on this for safety below; either change parse_number to `unsafe fn` or make this an unconditional assert.
```suggestion
pub fn parse_number(s: &[u8]) -> Option<(Number, usize)> {
    assert!(!s.is_empty());
```
```suggestion
pub unsafe fn parse_number(s: &[u8]) -> Option<(Number, usize)> {
    debug_assert!(!s.is_empty());
```

Reply:
[Fixed](https://github.com/Alexhuszagh/rust/blob/fb1a817fa13f774ce4da27fb9769db2779163030/library/core/src/num/dec2flt/mod.rs#L224-L232) by moving the sign-parsing logic outside `parse_number` and into `dec2flt`.

```rust
    if s.is_empty() {
        return Err(pfe_empty());
    }
    // SAFETY: cannot be empty
    let c = unsafe { s.first_unchecked() };
    let negative = c == b'-';
    if c == b'-' || c == b'+' {
        s = s.advance(1);
    }
```

Now, `parse_partial_number` (uses:

```rust
fn parse_partial_number(s: &[u8], negative: bool) -> Option<(Number, usize)> {
    let mut s = AsciiStr::new(s);
    let start = s;
    debug_assert!(!s.is_empty());

    // parse initial digits before dot
    let mut mantissa = 0_u64;
    let digits_start = s;
    try_parse_digits(&mut s, &mut mantissa);
    let mut n_digits = s.offset_from(&digits_start);

    ...
}
```